File: Data Structures & Algorithms/binary-tree-diameter/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */


class Solution {
    int max = 0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        int sum = length(root);
        return this.max;
    }


    public int length(TreeNode root){
        if(root == null) return 0;
        int a = length(root.left);
        int b = length(root.right);
        
   
        this.max = Math.max(this.max, a+b);
        if(a > b)
            return a + 1;
        return b+1;
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        
        return par(root, -1000);
    }

    int par(TreeNode* root, int parent){
        int s = 0;
        if(root == NULL) return 0;
        if(root->val >= parent){
            s++;
        } 

        int left = par(root->left, max(root->val, parent));
        int right = par(root->right, max(root->val, parent));

        return left + right+s;
    }

};
